File: sat/Sudoku.hs
{-|
  CNF - sudoku solver
-}

module Main (
-- * Main function
  -- *
) where
import System.IO
import System.Exit

type CNFindex = Integer -- Numeros del 1 al 9, no del 0 al 8 porque la var 0 no tiene negado (hint: -0 = 0)
type SudIndex = (Integer, Integer, Integer) -- Numeros del 1 al 9, no del 0 al 8 porque la var 0 no tiene negado (hint: -0 = 0)
type Clause   = [CNFindex]

-- helpers
sudIndex2CNFindex :: SudIndex -> CNFindex
sudIndex2CNFindex (i, j, k) = 81 * (i-1) + 9 * (j-1) + (k-1) + 1

cnfIndex2SudIndex :: CNFindex -> SudIndex
cnfIndex2SudIndex i = ((div (div (i-1) 9) 9) + 1, (mod (div (i-1) 9) 9) + 1, (mod (i-1) 9) + 1)

--
-- Static clauses
--

-- non-vacuity clauses
nonVacuity :: [Clause]
nonVacuity = nonVacuity' 1 1

nonVacuity' :: Integer -> Integer -> [Clause]
nonVacuity' i j
            | j > 9 = nonVacuity' (i+1) 0
            | i > 9 = []
            | otherwise = [(nonVacuity'' i j 1)] ++ (nonVacuity' i (j+1))

nonVacuity'' :: Integer -> Integer -> Integer -> Clause
nonVacuity'' i j k
             | k <= 9 = [sudIndex2CNFindex (i, j, k)] ++ nonVacuity'' i j (k+1)
             | otherwise = []

-- Uniqueness of each cell
uniqueness :: [Clause]
uniqueness = uniqueness' 1 1

uniqueness' :: Integer -> Integer -> [Clause]
uniqueness' i j
            | i > 9 = []
            | j > 9 = uniqueness' (i+1) 1
            | otherwise = (uniqueness'' i j 1 2) ++ (uniqueness' i (j+1))

uniqueness'' :: Integer -> Integer -> Integer -> Integer -> [Clause]
uniqueness'' i j d d'
             | d <= 9 && d' <= 9 = [[-sudIndex2CNFindex (i, j, d), -sudIndex2CNFindex (i, j, d')]] ++ uniqueness'' i j d (d'+1)
             | d > 9  = []
             | d' > 9 = uniqueness'' i j (d+1) (d+2)

-- Rows and colums
rowsRestrictions :: [Clause] -- Static clause list of sudoku rows restrictions
rowsRestrictions = rowsRestrictions' 1 1

rowsRestrictions' :: Integer -> Integer -> [Clause]
rowsRestrictions' i d
                  | i > 9 = []
                  | d > 9 = rowsRestrictions' (i+1) 1 -- next row
                  | otherwise = (rowsRestrictions'' i 1 2 d) ++ (rowsRestrictions' i (d+1))

rowsRestrictions'' :: Integer -> Integer -> Integer -> Integer -> [Clause]
rowsRestrictions'' i j j' d
                   | j <= 9 && j' <= 9 = [[-sudIndex2CNFindex (i, j, d), -sudIndex2CNFindex (i, j', d)]] ++ rowsRestrictions'' i j (j'+1) d
                   | j > 9  = []
                   | j' > 9 = rowsRestrictions'' i (j+1) (j+2) d


columnsRestrictions :: [Clause] -- Static clause list of sudoku columns restrictions
columnsRestrictions = columnsRestrictions' 1 1

columnsRestrictions' :: Integer -> Integer -> [Clause]
columnsRestrictions' j d
                     | j > 9 = []
                     | d > 9 = columnsRestrictions' (j+1) 1 -- next column
                     | otherwise = (columnsRestrictions'' 1 2 j d) ++ (columnsRestrictions' j (d+1))

columnsRestrictions'' :: Integer -> Integer -> Integer -> Integer -> [Clause]
columnsRestrictions'' i i' j d
                      | i <= 9 && i' <= 9 = [[-sudIndex2CNFindex (i, j, d), -sudIndex2CNFindex (i', j, d)]] ++ columnsRestrictions'' i (i'+1) j d
                      | i > 9  = []
                      | i' > 9 = columnsRestrictions'' (i+1) (i+2) j d


-- Blocks restrictions

File: sat/Sudoku.py
#!/usr/bin/python
import sys

# CNFIndex = Integer
# SudIndex = (Integer, Integer, Integer)
# Clause   = [CNFIndex]

# Helper functions
def sudIndex2CNFIndex(i, j, k):
    return 81 * (i-1) + 9 * (j-1) + (k-1) + 1

def cnfIndex2SudIndex(i):
    return (((i-1)/9/9) + 1, (((i-1)/9) % 9) + 1, ((i-1) % 9) + 1)

#
# Static Clauses
def non_vacuity():
    """
    returns [Clause]
    """
    clauses = []
    for i in range(1,10):
        clause = []
        for j in range(1,10):
            clause = [sudIndex2CNFIndex(i, j, k) for k in range(1,10)]
            clauses.append(clause)
    return clauses

def uniqueness():
    """
    """
    clauses = []
    for i in range(1,10):
        clauses_aux = []
        for j in range(1,10):
            clauses_aux = [[-sudIndex2CNFIndex(i,j,d1), -sudIndex2CNFIndex(i,j,d2)]
                           for d1 in range(1,9) for d2 in range(d1+1, 10)]
            clauses.extend(clauses_aux)
    return clauses

def rows_restrictions():
    """
    """
    clauses = []
    for i in range(1,10):
        clauses_aux = []
        for d in range(1,10):
            clauses_aux = [[-sudIndex2CNFIndex(i,j1,d), -sudIndex2CNFIndex(i,j2,d)]
                           for j1 in range(1,9) for j2 in range(j1+1,10)]
            clauses.extend(clauses_aux)
    return clauses

def columns_restrictions():
    """
    """
    clauses = []
    for j in range(1,10):
        clauses_aux = []
        for d in range(1,10):
            clauses_aux = [[-sudIndex2CNFIndex(i1,j,d), -sudIndex2CNFIndex(i2,j,d)]
                           for i1 in range(1,9) for i2 in range(i1+1,10)]
            clauses.extend(clauses_aux)
    return clauses

def blocks_restrictions():
    """
    """
    clauses = []
    for z in range(1,10):
        for i in range(0,3):
            for j in range(0,3):
                for x in range(1,4):
                    clauses_aux = [[-sudIndex2CNFIndex(3*i+x, 3*j+y, z),
                                    -sudIndex2CNFIndex(3*i+x, 3*j+k, z)]
                                   for y in range(1,3) for k in range(y+1, 4)]
                    clauses.extend(clauses_aux)

    for z in range(1,10):
        for i in range(0,3):
            for j in range(0,3):
                for l in range(1,4):
                    for y in range(1,4):
                        clauses_aux = [[-sudIndex2CNFIndex(3*i+x, 3*j+y, z),
                                        -sudIndex2CNFIndex(3*i+k, 3*j+l, z)]
                                       for x in range(1,3) for k in range(x+1, 4)]
                        clauses.extend(clauses_aux)
    return clauses

#
# Instance clauses
def instance_clauses(plain_sudoku):
    """
    """
    clauses = []
    for i in range(0, len(plain_sudoku)):
        if plain_sudoku[i] != '.':
            clauses.append([sudIndex2CNFIndex((int(i)/9)+1, (int(i)%9 +1),
                                              int(plain_sudoku[i]))])
    return clauses

                        
#
# Main function
def main():
    # check for the main argument and check it's length
    if len(sys.argv) != 2 or len(sys.argv[1]) != 81:
        sys.stderr.write(
"""
Modo de uso:
    > sudoku `sudoku_plain_representation`

donde `sudoku_plain_representation` es una cadena de caracteres
de longitud 81 descrita en el enunciado del proyecto

"""
                        )
    # Get static clauses
    clauses = non_vacuity() + uniqueness() + rows_restrictions() + \
              columns_restrictions() + blocks_restrictions()
    # Get instance clauses
    clauses.extend(instance_clauses(sys.argv[1]))

    # Print clauses in the zChaff's file
    zc_file = open('zcInput', 'w')
    zc_file.write('p cnf 729 %s\nc\nc Clauses:\nc\n' % len(clauses))
    for clause in clauses:
        zc_file.write('%s 0\n' % ' '.join([str(si) for si in clause]))
    zc_file.close()

if __name__ == "__main__": main()
